### PR TITLE
Num Shuffle Files : Gets the count of shuffle files pulled into memory for a filter condition 

### PR DESCRIPTION
**Pull request by joydeepbroy-zeotap**
This PR has two methods:
`getNumShuffleFiles` and `getShuffleFileMetadata`

First, get the count of shuffle files pulled into memory when a particular condition is provided in merge for a Delta Table.
The second one gets all the file metadata for the parquet files.

### CODE REVIEW

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala [hunk 1]
```diff
@@ -71,6 +73,111 @@ object DeltaHelpers {
...
+    deltaLog.withNewTransaction { deltaTxn =>
+      (deltaTxn.filterFiles(targetOnlyPredicates),
+        deltaTxn.filterFiles(minMaxOnlyExpressions),
+        deltaTxn.filterFiles(equalOnlyExpressions),
+        deltaTxn.filterFiles(otherExpressions),
+        deltaTxn.filterFiles(removedPredicates),
+        deltaLog.snapshot.filesWithStatsForScan(Nil),
+        unresolvedColumns)
+    }
+  }
```
Comment (joydeepbroy-zeotap) [inline src/main/scala/mrpowers/jodie/DeltaHelpers.scala]: I did not add any test cases for this method as the file metadata received is directly from Delta Table and I am not making any mutations to it. Additionally, please review if this method would be useful. I can remove it as well.

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala [hunk 2]
```diff
@@ -71,6 +73,111 @@ object DeltaHelpers {
...
+      case LessThan(_, _) => false
+      case GreaterThan(_, _) => false
+      case _ => true
+    })
+
+    val removedPredicates = splitExpressions.filterNot(_.references.subsetOf(targetPlan.outputSet))
+
+    (deltaLog, unresolvedColumns, targetOnlyPredicates, minMaxOnlyExpressions, equalOnlyExpressions, otherExpressions, removedPredicates)
+  }
+
```
Comment (joydeepbroy-zeotap) [inline src/main/scala/mrpowers/jodie/DeltaHelpers.scala:195]: this implementation is a little elaborate but it has been taken (at least in parts) from the Delta Merge code itself (just the relevant parts). That way it is more reliable and we know the result is exactly similar to what Delta Table would be fetching.

File: README.md [hunk 3]
```diff
@@ -438,6 +438,30 @@ DeltaHelpers.deltaNumRecordDistribution(path)
...
+The function `getNumShuffleFiles` gets the number of shuffle files (part files for parquet) that will be pulled into memory for a given filter condition. This is particularly useful in a Delta Merge operation where the number of shuffle files can be a bottleneck. 
+Running the merge condition through this method can give an idea about the amount of memory resources required to run the merge.
+For example, if the condition is "country = 'GBR' and age >= 30 and age <= 40 and firstname like '%Jo%' " and country is the partition column, 
+```scala
+DeltaHelpers.getNumShuffleFiles(path, "country = 'GBR' and age >= 30 and age <= 40 and firstname like '%Jo%' ")
+```
+
+then the output might look like 
+```scala
+(18, 100, 300, 600, 800, 800, List())
```
Comment (MrPowers) [inline README.md]: Should this output be a map that indicates what each value represents?

**Comment by MrPowers** [comment]
LGTM.  The list with multiple outputs might be hard to parse, but the functionality looks cool.  We should definitely write a blog post on this one.

File: README.md (hunk 3 above)
Comment (joydeepbroy-zeotap) [reply README.md]: I have changed the output to a MAP and updated the method comments. Please check if it looks good, then I will update README.md @MrPowers

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala [hunk 4]
```diff
@@ -71,6 +73,125 @@ object DeltaHelpers {
...
+    val targetPlan = deltaTable.toDF.queryExecution.analyzed
+    val resolvedExpression: Expression = spark.sessionState.analyzer.resolveExpressionByPlanOutput(expression, targetPlan, true)
+    val unresolvedColumns = if (!resolvedExpression.childrenResolved) {
+      resolvedExpression.references.filter(a => a match {
+        case b: UnresolvedAttribute => true
+        case _ => false
+      }).map(a => a.asInstanceOf[UnresolvedAttribute].sql).toSeq
+    } else Seq()
+
+    def splitConjunctivePredicates(condition: Expression): Seq[Expression] = {
```
Comment (brayanjuls) [inline src/main/scala/mrpowers/jodie/DeltaHelpers.scala:156]: I don't expect conditions to be massive enough to blow the stack but if we can make this function tail recursive it would be a good safeguard.

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala (hunk 2 above)
Comment (brayanjuls) [reply src/main/scala/mrpowers/jodie/DeltaHelpers.scala:195]: Just curious, could you point me to the delta API where you took part of this code?

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala (hunk 1 above)
Comment (brayanjuls) [reply src/main/scala/mrpowers/jodie/DeltaHelpers.scala]: I don't have any use case in mind but I think it may be of help for someone investigating the behavior of a query. My only advice here would be to explicitly add the return type in the method signature so it can serve as reference for someone that potentially would like to use this method.

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala [hunk 5]
```diff
@@ -71,6 +73,125 @@ object DeltaHelpers {
...
+    deltaLog.withNewTransaction { deltaTxn =>
+      Map(s"$OVERALL [ ${formatSQL(targetOnlyPredicates).getOrElse("Empty")} ]" ->
+        deltaTxn.filterFiles(targetOnlyPredicates).count(a => true),
+        s"$MIN_MAX [ ${formatSQL(minMaxOnlyExpressions).getOrElse("Empty")} ]" ->
+          deltaTxn.filterFiles(minMaxOnlyExpressions).count(a => true),
+        s"$EQUALS [ ${formatSQL(equalOnlyExpressions).getOrElse("Empty")} ]" ->
+          deltaTxn.filterFiles(equalOnlyExpressions).count(a => true),
+        s"$LEFT_OVER [ ${formatSQL(otherExpressions).getOrElse("Empty")} ]" ->
+          deltaTxn.filterFiles(otherExpressions).count(a => true),
+        s"$UNRESOLVED [ ${formatSQL(removedPredicates).getOrElse("Empty")} ]" ->
```
Comment (brayanjuls) [inline src/main/scala/mrpowers/jodie/DeltaHelpers.scala:121]: "UNRESOLVED"  and "UNRESOLVED_COLS" this mean we can have invalid expression and this method will still work, right ? i.e passing an invalid column name. My doubt is, would it be of help for users to get the number of shuffle files even when there are invalid conditions on their query?

**Review by brayanjuls — COMMENTED** [review]
Overall looks good, I just added a couple of comments.

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala (hunk 5 above)
Comment (joydeepbroy-zeotap) [reply src/main/scala/mrpowers/jodie/DeltaHelpers.scala:121]: I have kept it as a measure to double-check counts - for example, the condition that is erroneous vs. the one that does not have any condition - should ideally return the same answer as the count. If it does not then user needs to investigate

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala (hunk 4 above)
Comment (joydeepbroy-zeotap) [reply src/main/scala/mrpowers/jodie/DeltaHelpers.scala:156]: I have taken it from delta code base itself :- https://github.com/delta-io/delta/blob/8fe6ea6115f22b8341fce1917ee55974eb5f728d/connectors/standalone/src/main/scala/io/delta/standalone/internal/util/PartitionUtils.scala#L94

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala (hunk 2 above)
Comment (joydeepbroy-zeotap) [reply src/main/scala/mrpowers/jodie/DeltaHelpers.scala:195]: Basically, I debugged the delta merge command and looked into the data filters. Some classes and code to look at are

- [mergePlan](https://github.com/delta-io/delta/blob/8fe6ea6115f22b8341fce1917ee55974eb5f728d/spark/src/main/scala/io/delta/tables/DeltaMergeBuilder.scala#L304)
- [findTouchedFiles](https://github.com/delta-io/delta/blob/8fe6ea6115f22b8341fce1917ee55974eb5f728d/spark/src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala#L114)
- [dataSkippedFiles](https://github.com/delta-io/delta/blob/8fe6ea6115f22b8341fce1917ee55974eb5f728d/spark/src/main/scala/org/apache/spark/sql/delta/commands/merge/InsertOnlyMergeExecutor.scala#L95)

I have given links from the current github repo but debug using the delta 2.1.1 version with these lines of code

File: src/main/scala/mrpowers/jodie/DeltaHelpers.scala (hunk 5 above)
Comment (brayanjuls) [reply src/main/scala/mrpowers/jodie/DeltaHelpers.scala:121]: got it.

**Review by brayanjuls — APPROVED** [review]
LGTM.
